### PR TITLE
Proper expansion info for for loops and other compiler expanded language items

### PR DESCRIPTION
**Pull request by nrc**
r? @sfackler

### CODE REVIEW

**Comment by sfackler** [comment]
Why do we want expansion info here?

**Comment by nrc** [comment]
Because otherwise the spans are lies - we get a span for parts of the expansions of the for loop which are not what the span is for - i.e., a span for a type which is actually a value in the source. Furthermore, without expansion info you can't tell that the code has been expanded - a tool should assume that the expanded source is the same as the pre-expansion source, which it isn't.

File: src/libsyntax/ext/expand.rs [hunk 1]
```diff
@@ -304,20 +332,25 @@ pub fn expand_expr(e: P<ast::Expr>, fld: &mut MacroExpander) -> P<ast::Expr> {
...
                     span,
                     vec![fld.cx.stmt_let(span, false, result_ident, match_expr)],
-                    Some(fld.cx.expr_ident(span, result_ident))))
+                    Some(fld.cx.expr_ident(span, result_ident))));
+            fld.cx.bt_pop();
+            result
         }
 
         ast::ExprClosure(capture_clause, fn_decl, block) => {
+            push_compiler_expansion(fld, span, "closure expansion");
```
Comment (eddyb) [inline src/libsyntax/ext/expand.rs]: Came here from `git blame`, was this change accidental?
Closures are not expanded differently from free functions, so they shouldn't require custom expansion stacks.
All that does is pollute diagnostic output IMO.

Comment (nrc) [inline src/libsyntax/ext/expand.rs]: Yeah, probably wrong

**Pull request merged**